package/service/submission_service.go

```diff
@@ -13,17 +15,183 @@ type SubmissionService interface {
 	MarkSubmissionFailed(tx *gorm.DB, submissionId int64, errorMsg string) error
 	MarkSubmissionComplete(tx *gorm.DB, submissionId int64) error
 	MarkSubmissionProcessing(tx *gorm.DB, submissionId int64) error
+	CreateSubmission(tx *gorm.DB, taskId int64, userId int64, languageId int64, order int64) (int64, error)
 	CreateSubmissionResult(tx *gorm.DB, submissionId int64, responseMessage schemas.ResponseMessage) (int64, error)
+	GetAll(tx *gorm.DB, user schemas.UserSession, queryParams map[string][]string) ([]schemas.Submission, error)
+	GetById(tx *gorm.DB, submissionId int64, user schemas.UserSession) (schemas.Submission, error)
+	GetAllForUser(tx *gorm.DB, userId int64, user schemas.UserSession, queryParams map[string][]string) ([]schemas.Submission, error)
+	GetAllForGroup(tx *gorm.DB, groupId int64, user schemas.UserSession, queryParams map[string][]string) ([]schemas.Submission, error)
+	GetAllForTask(tx *gorm.DB, taskId int64, user schemas.UserSession, queryParams map[string][]string) ([]schemas.Submission, error)
 }
 
 type SubmissionServiceImpl struct {
 	submissionRepository       repository.SubmissionRepository
 	submissionResultRepository repository.SubmissionResultRepository
 	inputOutputRepository      repository.InputOutputRepository
 	testResultRepository       repository.TestResultRepository
+	userService                UserService
+	taskService                TaskService
+	languageService            LanguageService
 	logger                     *zap.SugaredLogger
 }
 
+var ErrPermissionDenied = errors.New("User is not allowed to view this submission")
+
+func (us *SubmissionServiceImpl) GetAll(tx *gorm.DB, user schemas.UserSession, queryParams map[string][]string) ([]schemas.Submission, error) {
+	submission_models := []models.Submission{}
+	var err error
+
+	switch user.Role {
+	case "admin":
+		submission_models, err = us.submissionRepository.GetAll(tx, queryParams)
+		break
+	case "student":
+		submission_models, err = us.submissionRepository.GetAllForStudent(tx, user.Id, queryParams)
+		break
+	case "teacher":
+		submission_models, err = us.submissionRepository.GetAllForTeacher(tx, user.Id, queryParams)
+		break
+	}
+
+	if err != nil {
+		us.logger.Errorf("Error getting all submissions: %v", err.Error())
+		return nil, err
+	}
+
+	var result []schemas.Submission
+	for _, submission_model := range submission_models {
+		result = append(result, *us.modelToSchema(&submission_model))
+	}
+
+	return result, nil
+}
+
+func (us *SubmissionServiceImpl) GetById(tx *gorm.DB, submissionId int64, user schemas.UserSession) (schemas.Submission, error) {
+	submission_model, err := us.submissionRepository.GetSubmission(tx, submissionId)
+	if err != nil {
+		us.logger.Errorf("Error getting submission: %v", err.Error())
+		return schemas.Submission{}, err
+	}
+
+	switch user.Role {
```

Comment:
https://go.dev/wiki/Switch#:~:text=Go's%20switch%20statements%20are%20pretty,the%20end%20of%20each%20case.